2.x: add Observable.switchMapSingle and switchMapSingleDelayError

davidmoten opened the pull request:
This is a new operator discussed in #4853. The issue refers to a goodly number of new operators which I'll do bit by bit as my time allows and to ensure review is not too daunting.

davidmoten committed:
add Observable.switchMapSingle

akarnokd commented on src/main/java/io/reactivex/internal/operators/observable/ObservableInternalHelper.java:
```diff
@@ -315,4 +315,31 @@ public boolean test(Notification<Object> t) throws Exception {
+                } else {
+                    single = Single.unsafeCreate(source);
+                }
+               return (Observable<R>) single.toObservable();
```
You could just use:

```java
return RxJavaPlugins.onAssembly(new SingleToObservable(mapper.apply(t)));
```

akarnokd commented on src/main/java/io/reactivex/Observable.java:
```diff
@@ -10867,6 +10867,65 @@ public final void subscribe(Observer<? super T> observer) {
+    @Experimental
+    @CheckReturnValue
+    @SchedulerSupport(SchedulerSupport.NONE)
+    public final <R> Observable<R> switchMapSingle(Function<? super T, ? extends SingleSource<? extends R>> mapper) {
```
`@NonNull`?

akarnokd commented on src/main/java/io/reactivex/Observable.java:
```diff
@@ -10867,6 +10867,65 @@ public final void subscribe(Observer<? super T> observer) {
+    @Experimental
+    @CheckReturnValue
+    @SchedulerSupport(SchedulerSupport.NONE)
+    public final <R> Observable<R> switchMapSingleDelayError(Function<? super T, ? extends SingleSource<? extends R>> mapper) {
```
`@NonNull Function`?

akarnokd commented on src/main/java/io/reactivex/internal/operators/observable/ObservableInternalHelper.java:
```diff
@@ -315,4 +315,31 @@ public boolean test(Notification<Object> t) throws Exception {
+    
+    private static <T, R> Function<T, Observable<R>> convertSingleMapperToObservableMapper(
+            final Function<? super T, ? extends SingleSource<? extends R>> mapper) {
+        return new Function<T, Observable<R>>() {
```
`ObjectHelper.requireNonNull(mapper, "mapper is null");`

akarnokd replied:
Please avoid anonymous inner classes.

akarnokd reviewed (commented):
Looks good but some changes are required.

davidmoten commented:
updated with suggested changes

davidmoten commented:
unrelated ci failure

```
io.reactivex.parallel.ParallelFlowableTest > parallelismAndPrefetchAsync FAILED
    java.lang.AssertionError: Value counts differ; Expected: 1048576, Actual: 814189 (latch = 1, values = 814189, errors = 0, completions = 0, timeout!, disposed!)
        at io.reactivex.observers.BaseTestConsumer.fail(BaseTestConsumer.java:163)
        at io.reactivex.observers.BaseTestConsumer.assertValueCount(BaseTestConsumer.java:462)
        at io.reactivex.parallel.ParallelFlowableTest.parallelismAndPrefetchAsync(ParallelFlowableTest.java:741)
```

akarnokd commented:
> unrelated ci failure

Tracking via #5154. Could be due to low timeout settings and Travis overload. Just rerun the build next time; I did it just now.

vanniktech commented on src/test/java/io/reactivex/internal/operators/observable/ObservableSwitchTest.java:
```diff
@@ -622,8 +622,65 @@ public void switchMapInnerCancelled() {
+    @Test
+    public void switchMapSingleFunctionDoesntReturnSingle() {
+        Observable.just(0)
+        .switchMapSingle(new Function<Object, SingleSource<Integer>>() {
```
I remember that with flatMapSingle there was in the initial draft a few errors that errors weren't propagated correctly. Should not those cases also be tested that in the future once the operators receive custom implementations everything still works correctly?

davidmoten replied:
I don't think so. At the moment those cases have full coverage. If we make custom implementations then a whole suite of tests need to be added just get coverage back up again and it will be an obvious timesaver to duplicate the tests you refer to.

vanniktech commented on src/main/java/io/reactivex/internal/operators/observable/ObservableInternalHelper.java:
```diff
@@ -315,4 +318,34 @@ public boolean test(Notification<Object> t) throws Exception {
+        
+        @Override
+        public Observable<R> apply(T t) throws Exception {
+            return RxJavaPlugins.onAssembly(new SingleToObservable<R>(mapper.apply(t)));
```
check that `mapper.apply(t)` does not return null?

akarnokd replied:
Indeed, this should do it:

```java
return RxJavaPlugins.onAssembly(new SingleToObservable<R>(
    ObjectHelper.requireNonNull(mapper.apply(t), "The mapper returned a null value")));
```

davidmoten replied:
thanks will do

davidmoten commented:
updated with null check on mapper call and added unit tests for null mapper and null mapper call result. Not proposing to duplicate all tests. @akarnokd what would you like?

akarnokd commented:
Don't duplicate tests. When the specialized implementation happens, that will ask for proper coverage by itself.
